x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/license/PutLicenseResponse.java

```diff
@@ -0,0 +1,209 @@
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class PutLicenseResponse extends AcknowledgedResponse {
+
+    private static final ConstructingObjectParser<PutLicenseResponse, Void> PARSER = new ConstructingObjectParser<>(
+        "put_license_response", true, (a, v) -> {
+        boolean acknowledged = (Boolean) a[0];
+        LicensesStatus licensesStatus = LicensesStatus.fromString((String) a[1]);
+        @SuppressWarnings("unchecked") Tuple<String, Map<String, String[]>> acknowledgements = (Tuple<String, Map<String, String[]>>) a[2];
+        if (acknowledgements == null) {
+            return new PutLicenseResponse(acknowledged, licensesStatus);
+        } else {
+            return new PutLicenseResponse(acknowledged, licensesStatus, acknowledgements.v1(), acknowledgements.v2());
+        }
+
+    });
+
+    static {
+        PARSER.declareBoolean(constructorArg(), new ParseField("acknowledged"));
+        PARSER.declareString(constructorArg(), new ParseField("license_status"));
+        PARSER.declareObject(optionalConstructorArg(), (parser, v) -> {
+                Map<String, String[]> acknowledgeMessages = new HashMap<>();
+                String message = null;
+                XContentParser.Token token;
+                String currentFieldName = null;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        currentFieldName = parser.currentName();
+                    } else {
+                        if (currentFieldName == null) {
+                            throw new XContentParseException(parser.getTokenLocation(), "expected message header or acknowledgement");
+                        }
+                        if ("message".equals(currentFieldName)) {
+                            if (token != XContentParser.Token.VALUE_STRING) {
+                                throw new XContentParseException(parser.getTokenLocation(), "unexpected message header type");
+                            }
+                            message = parser.text();
+                        } else {
+                            if (token != XContentParser.Token.START_ARRAY) {
+                                throw new XContentParseException(parser.getTokenLocation(), "unexpected acknowledgement type");
+                            }
+                            List<String> acknowledgeMessagesList = new ArrayList<>();
+                            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                                if (token != XContentParser.Token.VALUE_STRING) {
+                                    throw new XContentParseException(parser.getTokenLocation(), "unexpected acknowledgement text");
+                                }
+                                acknowledgeMessagesList.add(parser.text());
```

Comment:
What if this contains an object? I think calling `nextToken` in that case will move into the object which isn't quite what you want. I *think* we want to handle that kind of evil stuff in these responses because it gives us the flexibility to make those changes eventually.

Reply:
Not sure I follow. If it contains an object, it will fail 8 lines above this one.

Reply:
oh, nm, it seems like this was answered by @imotov